src/boot/efi/shim.c

```diff
@@ -0,0 +1,255 @@
+/*
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * Port to systemd-boot
+ * Copyright 2017 Max Resch <resch.max@gmail.com>
+ *
+ * Security Policy Handling
+ * Copyright 2012 <James.Bottomley@HansenPartnership.com>
+ * https://github.com/mjg59/efitools
+ */
+
+#include <efi.h>
+#include <efilib.h>
+
+#include "util.h"
+#include "shim.h"
+
+static EFI_GUID SECURITY_PROTOCOL_GUID = { 0xA46423E3, 0x4617, 0x49f1, {0xB9, 0xFF, 0xD1, 0xBF, 0xA9, 0x11, 0x58, 0x39 } };
+static EFI_GUID SECURITY2_PROTOCOL_GUID = { 0x94ab2f58, 0x1438, 0x4ef1, {0x91, 0x52, 0x18, 0x94, 0x1a, 0x3a, 0x0e, 0x68 } };
+static EFI_GUID SIMPLE_FS_PROTOCOL = SIMPLE_FILE_SYSTEM_PROTOCOL;
+static EFI_GUID GLOBAL_GUID = EFI_GLOBAL_VARIABLE;
+
+BOOLEAN shim_loaded(void) {
+        SHIM_LOCK   *shim_lock;
+        EFI_GUID    ShimLockGuid = SHIM_LOCK_GUID;
+
+        return (uefi_call_wrapper(BS->LocateProtocol, 3, &ShimLockGuid, NULL, (VOID**) &shim_lock) == EFI_SUCCESS);
+}
+
+static BOOLEAN shim_validate(VOID *data, UINT32 size) {
+        SHIM_LOCK   *shim_lock;
+        EFI_GUID    ShimLockGuid = SHIM_LOCK_GUID;
+        EFI_STATUS  result;
+
+        if ((data != NULL) && (uefi_call_wrapper(BS->LocateProtocol, 3, &ShimLockGuid, NULL, (VOID**) &shim_lock) == EFI_SUCCESS)) {
```

Comment:
I think it would be nicer, to check the data != NULL in a separate if check and exit early, so that we don't have to indent the block in that case here...

i.e. just do:

```c
if (!data)
        return FALSE
```

early.